### PR TITLE
Add commands to change the worker key

### PR DESCRIPTION
**Pull request by Stebalien**

### CODE REVIEW

File: cmd/lotus-storage-miner/actor.go [hunk 1]
```diff
@@ -701,3 +701,196 @@ var actorSetOwnerCmd = &cli.Command{
...
+		}
+
+		return nil
+	},
+}
+
+var actorConfirmChangeWorker = &cli.Command{
+	Name:      "confirm-change-worker",
+	Usage:     "Confirm a worker address change",
+	ArgsUsage: "[address]",
```
Comment (Stebalien) [inline cmd/lotus-storage-miner/actor.go:812]: I could drop this flag, but it's nice to check twice.

File: cmd/lotus-storage-miner/actor.go [hunk 2]
```diff
@@ -701,3 +701,196 @@ var actorSetOwnerCmd = &cli.Command{
...
+		na, err := address.NewFromString(cctx.Args().First())
+		if err != nil {
+			return err
+		}
+
+		newAddr, err := api.StateLookupID(ctx, na, types.EmptyTSK)
+		if err != nil {
+			return err
+		}
+
```
Comment (Stebalien) [inline cmd/lotus-storage-miner/actor.go:749]: Should we check to see if we have the key in our wallet?

File: cmd/lotus-storage-miner/actor.go [hunk 3]
```diff
@@ -701,3 +701,196 @@ var actorSetOwnerCmd = &cli.Command{
...
+			Name:  "really-do-it",
+			Usage: "Actually send transaction performing the action",
+			Value: false,
+		},
+	},
+	Action: func(cctx *cli.Context) error {
+		if !cctx.Bool("really-do-it") {
+			fmt.Println("Pass --really-do-it to actually execute this action")
+			return nil
+		}
```
Comment (Stebalien) [inline cmd/lotus-storage-miner/actor.go:824]: Probably not necessary for this command?

File: node/modules/storageminer.go [hunk 4]
```diff
@@ -186,22 +186,12 @@ func StorageMiner(fc config.MinerFeeConfig) func(params StorageMinerParams) (*st
...
-		mi, err := api.StateMinerInfo(ctx, maddr, types.EmptyTSK)
+		fps, err := storage.NewWindowedPoStScheduler(api, fc, sealer, sealer, j, maddr)
 		if err != nil {
 			return nil, err
 		}
 
-		worker, err := api.StateAccountKey(ctx, mi.Worker, types.EmptyTSK)
-		if err != nil {
-			return nil, err
-		}
```
Comment (Stebalien) [inline node/modules/storageminer.go:197]: I assume we just needed to do this to perform the sanity checks inside the miner. I'm now _only_ resolving the worker key there: https://github.com/filecoin-project/lotus/pull/4513/files#diff-e5e6c18c093f7eb4f4d0bb2e19388a831f34ded5ceeff4737091f6dc686aafe6R181-R184

File: storage/wdpost_run.go [hunk 5]
```diff
@@ -752,21 +755,17 @@ func (s *WindowPoStScheduler) submitPost(ctx context.Context, proof *miner.Submi
...
 
-func (s *WindowPoStScheduler) setSender(ctx context.Context, msg *types.Message, spec *api.MessageSendSpec) {
+func (s *WindowPoStScheduler) setSender(ctx context.Context, msg *types.Message, spec *api.MessageSendSpec) error {
 	mi, err := s.api.StateMinerInfo(ctx, s.actor, types.EmptyTSK)
 	if err != nil {
-		log.Errorw("error getting miner info", "error", err)
-
-		// better than just failing
-		msg.From = s.worker
-		return
```
Comment (Stebalien) [inline storage/wdpost_run.go:762]: So... I _could_ try playing some kind of caching game, but if we can't even get the worker key, something is very wrong.

**Comment by Stebalien** [comment]
Looks like real test failures, still need to figure out what I broke...

**Comment by Stebalien** [comment]
Ok... fixed.

File: cmd/lotus-storage-miner/actor.go (hunk 2 above)
Comment (Stebalien) [reply cmd/lotus-storage-miner/actor.go:751]: Hm. My answer is "no"? We might want to run this command on a node with only the owner key, right?

Comment (whyrusleeping) [reply cmd/lotus-storage-miner/actor.go:751]: Yeah, I think thats right. We don't necessarily need the key on the node running this command

File: cmd/lotus-storage-miner/actor.go [hunk 6]
```diff
@@ -698,3 +700,221 @@ var actorSetOwnerCmd = &cli.Command{
...
+			Name:  "really-do-it",
+			Usage: "Actually send transaction performing the action",
+			Value: false,
+		},
+	},
+	Action: func(cctx *cli.Context) error {
+		if !cctx.Bool("really-do-it") {
+			fmt.Fprintln(cctx.App.Writer, "Pass --really-do-it to actually execute this action")
+			return nil
+		}
```
Comment (magik6k) [inline cmd/lotus-storage-miner/actor.go:719]: I'd put this after sanity checks, so running this command without that flag is actually useful

File: cmd/lotus-storage-miner/actor.go [hunk 7]
```diff
@@ -698,3 +700,221 @@ var actorSetOwnerCmd = &cli.Command{
...
```
Comment (magik6k) [inline cmd/lotus-storage-miner/actor.go:834]: Same as above

**Review by magik6k — COMMENTED** [review]
Looks like this may even work, would be really good to test on some devnet / in pond first

Just 1.5 nits

File: api/test/deals.go [hunk 8]
```diff
@@ -34,18 +31,7 @@ import (
...
-	Done:        func(bool, abi.ChainEpoch, error) {},
-}
-
-func init() {
-	logging.SetAllLoggers(logging.LevelInfo)
-	build.InsecurePoStValidation = true
-}
-
 func TestDealFlow(t *testing.T, b APIBuilder, blocktime time.Duration, carExport, fastRet bool) {
-	_ = os.Setenv("BELLMAN_NO_GPU", "1")
```
Comment (whyrusleeping) [inline api/test/deals.go:48]: why remove all this?

Comment (whyrusleeping) [reply api/test/deals.go:48]: oh, i see refactored it into init

Comment (Stebalien) [reply api/test/deals.go:48]: I'm setting it once in an `init` function. Setting globals in different places was making it difficult to tell which globals applied to which tests.

Comment (Stebalien) [reply api/test/deals.go]: Merged in https://github.com/filecoin-project/lotus/pull/4538 (broken out of this PR).

**Review by whyrusleeping — APPROVED** [review]
LGTM

**Comment by Stebalien** [comment]
Fixed and rebased.

File: storage/wdpost_run.go [hunk 9]
```diff
@@ -750,33 +753,18 @@ func (s *WindowPoStScheduler) submitPost(ctx context.Context, proof *miner.Submi
...
-
-		msg.From = s.worker
-		return
-	}
-	// worker should keep sync with state on chain, if not, will met error when estimate msg gas.
-	if s.worker != worker {
-		s.worker = worker
-		msg.From = s.worker
+		return xerrors.Errorf("error getting miner info: %w", err)
 	}
```
Comment (Stebalien) [inline storage/wdpost_run.go:760]: My original patch didn't bother caching (see above comment) but this was added in by a subsequent commit on master.  I can add this back if desired, but it seems unnecessary and storing the key seems error prone.

**Review by whyrusleeping — APPROVED** [review]

**Review by Kubuxu — APPROVED** [review]

**Pull request merged**